File: src/Tests/Behavioral/IotaEnum/IotaEnum.cs
namespace go;

using fmt = fmt_package;
using @unsafe = unsafe_package;

partial class main_package {

[GoType] partial struct Type {
    public uintptr Size_;
    public uintptr PtrBytes;
    public uint32 Hash;
    public TFlag TFlag;
    public uint8 Align_;
    public uint8 FieldAlign_;
    public ΔKind Kind_;
    public Func<@unsafe.Pointer, @unsafe.Pointer, bool> Equal;
    public ж<byte> GCData;
    public NameOff Str;
    public TypeOff PtrToThis;
}

[GoType("num:uint8")] partial struct ΔKind;

public static readonly ΔKind Invalid = /* iota */ 0;
public static readonly ΔKind Bool = 1;
public static readonly ΔKind Int = 2;
public static readonly ΔKind Int8 = 3;
public static readonly ΔKind Int16 = 4;
public static readonly ΔKind Int32 = 5;
public static readonly ΔKind Int64 = 6;
public static readonly ΔKind Uint = 7;
public static readonly ΔKind Uint8 = 8;
public static readonly ΔKind Uint16 = 9;
public static readonly ΔKind Uint32 = 10;
public static readonly ΔKind Uint64 = 11;
public static readonly ΔKind Uintptr = 12;
public static readonly ΔKind Float32 = 13;
public static readonly ΔKind Float64 = 14;
public static readonly ΔKind Complex64 = 15;
public static readonly ΔKind Complex128 = 16;
public static readonly ΔKind Array = 17;
public static readonly ΔKind Chan = 18;
public static readonly ΔKind Func = 19;
public static readonly ΔKind Interface = 20;
public static readonly ΔKind Map = 21;
public static readonly ΔKind Pointer = 22;
public static readonly ΔKind Slice = 23;
public static readonly ΔKind ΔString = 24;
public static readonly ΔKind Struct = 25;
public static readonly ΔKind UnsafePointer = 26;

public static @string String(this ΔKind k) {
    if ((nint)(uint8)k < len(kindNames)) {
        return kindNames[k];
    }
    return kindNames[0];
}

internal static slice<@string> kindNames = new golib.SparseArray<@string>{
    [Invalid] = "invalid"u8,
    [Bool] = "bool"u8,
    [Int] = "int"u8,
    [Int8] = "int8"u8,
    [Int16] = "int16"u8,
    [Int32] = "int32"u8,
    [Int64] = "int64"u8,
    [Uint] = "uint"u8,
    [Uint8] = "uint8"u8,
    [Uint16] = "uint16"u8,
    [Uint32] = "uint32"u8,
    [Uint64] = "uint64"u8,
    [Uintptr] = "uintptr"u8,
    [Float32] = "float32"u8,
    [Float64] = "float64"u8,
    [Complex64] = "complex64"u8,
    [Complex128] = "complex128"u8,
    [Array] = "array"u8,
    [Chan] = "chan"u8,
    [Func] = "func"u8,
    [Interface] = "interface"u8,
    [Map] = "map"u8,
    [Pointer] = "ptr"u8,
    [Slice] = "slice"u8,
    [ΔString] = "string"u8,
    [Struct] = "struct"u8,
    [UnsafePointer] = "unsafe.Pointer"u8
}.slice();

[GoType("num:int32")] partial struct NameOff;

[GoType("num:int32")] partial struct TypeOff;

[GoType("num:int32")] partial struct TextOff;

public static readonly ΔKind KindDirectIface = /* 1 << 5 */ 32;
public static readonly ΔKind KindGCProg = /* 1 << 6 */ 64;
public static readonly ΔKind KindMask = /* (1 << 5) - 1 */ 31;

[GoType("num:uint8")] partial struct TFlag;

public static readonly TFlag TFlagUncommon = /* 1 << 0 */ 1;
public static readonly TFlag TFlagExtraStar = /* 1 << 1 */ 2;
public static readonly TFlag TFlagNamed = /* 1 << 2 */ 4;
public static readonly TFlag TFlagRegularMemory = /* 1 << 3 */ 8;
public static readonly TFlag TFlagUnrolledBitmap = /* 1 << 4 */ 16;

public static @unsafe.Pointer NoEscape(@unsafe.Pointer p) {
    var x = (uintptr)p;
    return (@unsafe.Pointer)((uintptr)(x ^ 0));
}

[GoType] partial struct EmptyInterface {
    public ж<Type> Type;
    public @unsafe.Pointer Data;
}

public static ж<Type> TypeOf(any a) {
    var eface = ~(ж<EmptyInterface>)(uintptr)(new @unsafe.Pointer(Ꮡ(a)));
    return eface.Type;
}

public static ж<Type> TypeFor<T>() {
    T v = default!;
    {
        var t = TypeOf(v); if (t != nil) {
            return t;
        }
    }
    return TypeOf(((ж<T>)nil)).Elem();
}

[GoRecv] public static ΔKind Kind(this ref Type t) {
    return (ΔKind)(t.Kind_ & KindMask);
}

[GoRecv] public static bool HasName(this ref Type t) {
    return (TFlag)(t.TFlag & TFlagNamed) != 0;
}

public static ж<Type> Elem(this ж<Type> Ꮡt) {
    ref var t = ref Ꮡt.Value;

    var exprᴛ1 = t.Kind();
    if (exprᴛ1 == Array) {
        var tt = (ж<ΔArrayType>)(uintptr)(@unsafe.Pointer.FromRef(ref t));
        return (~tt).Elem;
    }
    if (exprᴛ1 == Map) {
        var tt = (ж<ΔMapType>)(uintptr)(@unsafe.Pointer.FromRef(ref t));
        return (~tt).Elem;
    }

    return default!;
}

[GoType] partial struct ΔMapType {
    public partial ref Type Type { get; }
    public ж<Type> Key;
    public ж<Type> Elem;
    public ж<Type> Bucket;
    public Func<@unsafe.Pointer, uintptr, uintptr> Hasher;
    public uint8 KeySize;
    public uint8 ValueSize;
    public uint16 BucketSize;
    public uint32 Flags;
}

public static ж<ΔMapType> MapType(this ж<Type> Ꮡt) {
    ref var t = ref Ꮡt.Value;

    if (t.Kind() != Map) {
        return default!;
    }
    return (ж<ΔMapType>)(uintptr)(@unsafe.Pointer.FromRef(ref t));
}

[GoType] partial struct ΔArrayType {
    public partial ref Type Type { get; }
    public ж<Type> Elem;
    public ж<Type> Slice;
    public uintptr Len;
}

public static ж<ΔArrayType> ArrayType(this ж<Type> Ꮡt) {
    ref var t = ref Ꮡt.Value;

    if (t.Kind() != Array) {
        return default!;
    }
    return (ж<ΔArrayType>)(uintptr)(@unsafe.Pointer.FromRef(ref t));
}

internal const nint seqFirst = iota;
internal const nint seqSecond = 1;
internal const nint seqAgain = /* iota */ 2;

internal const int64 wideFirst = /* iota */ 0;
internal const int64 wideNext = 1;

internal static readonly UntypedInt rawZero = iota;
internal static readonly UntypedInt rawOne = /* iota */ 1;

internal static @string stateMachine(nint step) {
    const nint stateInit = iota;
    const nint stateDict = 1;
    const int64 wideLocal = /* iota */ 0;
    int64 offset = wideLocal;
    var exprᴛ1 = step;
    if (exprᴛ1 == stateInit) {
        return fmt.Sprintf("init[%d]"u8, offset);
    }
    if (exprᴛ1 == stateDict) {
        return fmt.Sprintf("dict[%d]"u8, offset);
    }

    return "unknown"u8;
}

internal static void Main() {
    fmt.Printf("Slice Kind Value: %s [%d]\n"u8, Slice.String(), (nint)(uint8)Slice);
    fmt.Println(seqFirst, seqSecond, seqAgain);
    fmt.Println(wideFirst, wideNext);
    fmt.Println(rawZero, rawOne);
    fmt.Println(stateMachine(0));
    fmt.Println(stateMachine(1));
    fmt.Println(stateMachine(2));
}

} // end main_package

File: src/Tests/Behavioral/IotaEnum/IotaEnum.go
package main

import (
	"fmt"
	"unsafe"
)

type Type struct {
	Size_       uintptr
	PtrBytes    uintptr
	Hash        uint32
	TFlag       TFlag
	Align_      uint8
	FieldAlign_ uint8
	Kind_       Kind
	Equal       func(unsafe.Pointer, unsafe.Pointer) bool
	GCData      *byte
	Str         NameOff
	PtrToThis   TypeOff
}

type Kind uint8

const (
	Invalid Kind = iota
	Bool
	Int
	Int8
	Int16
	Int32
	Int64
	Uint
	Uint8
	Uint16
	Uint32
	Uint64
	Uintptr
	Float32
	Float64
	Complex64
	Complex128
	Array
	Chan
	Func
	Interface
	Map
	Pointer
	Slice
	String
	Struct
	UnsafePointer
)

// String returns the name of k.
func (k Kind) String() string {
	if int(k) < len(kindNames) {
		return kindNames[k]
	}
	return kindNames[0]
}

var kindNames = []string{
	Invalid:       "invalid",
	Bool:          "bool",
	Int:           "int",
	Int8:          "int8",
	Int16:         "int16",
	Int32:         "int32",
	Int64:         "int64",
	Uint:          "uint",
	Uint8:         "uint8",
	Uint16:        "uint16",
	Uint32:        "uint32",
	Uint64:        "uint64",
	Uintptr:       "uintptr",
	Float32:       "float32",
	Float64:       "float64",
	Complex64:     "complex64",
	Complex128:    "complex128",
	Array:         "array",
	Chan:          "chan",
	Func:          "func",
	Interface:     "interface",
	Map:           "map",
	Pointer:       "ptr",
	Slice:         "slice",
	String:        "string",
	Struct:        "struct",
	UnsafePointer: "unsafe.Pointer",
}

type NameOff int32
type TypeOff int32
type TextOff int32

const (
	KindDirectIface Kind = 1 << 5
	KindGCProg      Kind = 1 << 6
	KindMask        Kind = (1 << 5) - 1
)

type TFlag uint8

const (
	TFlagUncommon       TFlag = 1 << 0
	TFlagExtraStar      TFlag = 1 << 1
	TFlagNamed          TFlag = 1 << 2
	TFlagRegularMemory  TFlag = 1 << 3
	TFlagUnrolledBitmap TFlag = 1 << 4
)

func NoEscape(p unsafe.Pointer) unsafe.Pointer {
	x := uintptr(p)
	return unsafe.Pointer(x ^ 0)
}

type EmptyInterface struct {
	Type *Type
	Data unsafe.Pointer
}

func TypeOf(a any) *Type {
	eface := *(*EmptyInterface)(unsafe.Pointer(&a))
	return (*Type)(NoEscape(unsafe.Pointer(eface.Type)))
}

func TypeFor[T any]() *Type {
	var v T
	if t := TypeOf(v); t != nil {
		return t
	}
	return TypeOf((*T)(nil)).Elem()
}

func (t *Type) Kind() Kind { return t.Kind_ & KindMask }

func (t *Type) HasName() bool {
	return t.TFlag&TFlagNamed != 0
}

func (t *Type) Elem() *Type {
	switch t.Kind() {
	case Array:
		tt := (*ArrayType)(unsafe.Pointer(t))
		return tt.Elem
	case Map:
		tt := (*MapType)(unsafe.Pointer(t))
		return tt.Elem
	}
	return nil
}

type MapType struct {
	Type
	Key        *Type
	Elem       *Type
	Bucket     *Type
	Hasher     func(unsafe.Pointer, uintptr) uintptr
	KeySize    uint8
	ValueSize  uint8
	BucketSize uint16
	Flags      uint32
}

func (t *Type) MapType() *MapType {
	if t.Kind() != Map {
		return nil
	}
	return (*MapType)(unsafe.Pointer(t))
}

type ArrayType struct {
	Type
	Elem  *Type // array element type
	Slice *Type // slice type
	Len   uintptr
}

func (t *Type) ArrayType() *ArrayType {
	if t.Kind() != Array {
		return nil
	}
	return (*ArrayType)(unsafe.Pointer(t))
}

// Bare-iota emission rule: golib's builtin declares `const nint iota = 0`. A const
// initialized by exactly the builtin `iota` emits it bare only when the emitted C# type
// matches golib's constant (nint — Go int, explicit or tightened) AND the folded
// group-position value is 0; every other type (named wrappers like Kind above, other
// widths like int64) and every later group position keeps the folded `/* iota */ N` form.
const (
	seqFirst  int = iota // position 0 at int -> `const nint seqFirst = iota`
	seqSecond            // implicit repetition -> folded 1
	seqAgain  int = iota // later position, explicit iota -> `/* iota */ 2`
)

const (
	wideFirst int64 = iota // position 0 but int64 != nint -> `/* iota */ 0`
	wideNext
)

const (
	rawZero = iota // untyped at package level (never tightens) -> `UntypedInt rawZero = iota`
	rawOne  = iota // later position: value 1 is not golib's iota -> `/* iota */ 1`
)

// stateMachine mirrors compress/flate's function-local const group: every use of the
// state consts below is a concrete int context, so the group TIGHTENS to nint and
// position 0 emits bare `iota`; wideLocal's only use is int64, so it tightens to a
// non-nint type and keeps the folded form.
func stateMachine(step int) string {
	const (
		stateInit = iota // tightens to nint -> `const nint stateInit = iota`
		stateDict
	)

	const wideLocal = iota // tightens to int64 -> `const int64 wideLocal = /* iota */ 0`
	var offset int64 = wideLocal

	switch step {
	case stateInit:
		return fmt.Sprintf("init[%d]", offset)
	case stateDict:
		return fmt.Sprintf("dict[%d]", offset)
	}
	return "unknown"
}

func main() {
	fmt.Printf("Slice Kind Value: %s [%d]\n", Slice.String(), int(Slice))
	fmt.Println(seqFirst, seqSecond, seqAgain)
	fmt.Println(wideFirst, wideNext)
	fmt.Println(rawZero, rawOne)
	fmt.Println(stateMachine(0))
	fmt.Println(stateMachine(1))
	fmt.Println(stateMachine(2))
}
